Added backend support for allowed post types

rhysyngsun opened the pull request:
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [ ] Testing
  - [x] Code is tested
  - [ ] Changes have been manually tested

#### What are the relevant tickets?
Part of #1495 

#### What's this PR do?
This updates the backend to accept a new configuration parameter for channels `allowed_posts_types`, which is a list of allowed types. The reason for this is to support the new "article" post type, while still integrating with native reddit post types and continuing to support the now-deprecated `link_type` field on channels.

#### How should this be manually tested?
- Test creating a channel in MM
- Test creating a channel in the discussions ui
- Test updating channel allowed post types in the ui
- Test that you can still create the allowed post types in those channels and can't create the ones that are disallowed

mbertrand commented on channels/tasks.py:
```diff
@@ -131,3 +138,73 @@ def populate_subscriptions_and_roles(self):
+        if channel.allowed_post_types:
+            continue
+
+        subreddit = client.get_subreddit(channel.display_name)
```
Use `channel.name` instead?  When this task runs via the `backpopulate_post_tasks` command, it throws an 
`AttributeError: 'Channel' object has no attribute 'display_name'`

rhysyngsun commented on channels/tasks.py:
```diff
@@ -131,3 +138,73 @@ def populate_subscriptions_and_roles(self):
+        else:
+            post.post_type = LINK_TYPE_SELF
+
+        post.savE()
```
Probably more critically, a test would've caught this :woman_facepalming:

rhysyngsun commented:
@mbertrand addressed your feedback and added tests around the code. I also fixed the name generation on `ChannelFactory.create_batch()` to guarantee uniqueness since I ran into that.

mbertrand commented on channels/tasks.py:
```diff
@@ -131,3 +139,73 @@ def populate_subscriptions_and_roles(self):
+            )
+        ]
+        + [
+            populate_channel_post_type.si(ids)
```
When run from `backpopulate_post_types` command, this prevents all existing channels from allowing creation of any new article posts (via `channels.api.get_allowed_post_types_from_link_type`).  Is this by design, so that channels have to be manually set to allow articles afterward?

rhysyngsun replied:
Yes, that's intentional, because we want to deploy this with no functional or configuration change to production.
